### ISSUE TITLE
`datalad run` no longer relays exit code

### ISSUE THREAD

**Issue by mih** [datalad/datalad#7504]
This used to be a thing, and the handbook also documents it still. Somewhere in the 0.16 series it got lost.

```
❯ chronic pip install datalad==0.15.6
❯ dl run "bash -c 'exit 3'"
[INFO   ] == Command start (output follows) ===== 
[INFO   ] == Command exit (modification check follows) ===== 
[INFO   ] The command had a non-zero exit code. If this is expected, you can save the changes with 'datalad save -d . -r -F .git/COMMIT_EDITMSG' 
CommandError: 'bash -c 'exit 3'' failed with exitcode 3 under /tmp/some
❯ echo $?
3

❯ chronic pip install datalad==0.16.4
❯ dl run "bash -c 'exit 3'"
[WARNING] Requested extension 'next' is not available 
[INFO   ] == Command start (output follows) ===== 
[INFO   ] == Command exit (modification check follows) ===== 
[INFO   ] The command had a non-zero exit code. If this is expected, you can save the changes with 'datalad save -d . -r -F .git/COMMIT_EDITMSG' 
run(error): /tmp/some (dataset) [bash -c 'exit 3']
❯ echo $?
1
```

I think this was a useful feature, and we should restore it.

I believe this is caused by a change in the handling. Now we get an error result, rather than an exception handler at the very top.

That being said, the result itself contains everything needed to set a matching exit code:

```
❯ dl -f json_pp run "bash -c 'exit 3'"
[WARNING] Requested extension 'next' is not available 
[INFO   ] == Command start (output follows) ===== 
[INFO   ] == Command exit (modification check follows) ===== 
{
  "action": "run",
  "exception": "CommandError: 'bash -c 'exit 3'' failed with exitcode 3 under /tmp/some",
  "exception_traceback": "[run.py:_execute_command:701,runner.py:run:242]",
  "exit_code": 3,
  "message": "bash -c 'exit 3'",
  "msg_path": "/tmp/some/.git/COMMIT_EDITMSG",
  "path": "/tmp/some",
  "run_info": {
    "chain": [],
    "cmd": "bash -c 'exit 3'",
    "dsid": "9fdb9a12-d3f5-4b19-910d-ad1b617a912b",
    "exit": 3,
    "extra_inputs": [],
    "inputs": [],
    "outputs": [],
    "pwd": "."
  },
  "status": "error",
  "type": "dataset"
}
```

**Comment by mih** [comment]
The following patch would restore this behavior:

```diff
diff --git a/datalad/cli/main.py b/datalad/cli/main.py
index bb68e46c8..1106b99db 100644
--- a/datalad/cli/main.py
+++ b/datalad/cli/main.py
@@ -208,6 +208,14 @@ def _run_with_exception_handler(cmdlineargs):
             # in general we do not want to see the error again, but
             # present in debug output
             lgr.debug('could not perform all requested actions: %s', ce)
+            # fish for an exit code. If any of the "failed" results
+            # is caused by a command error, it will come with an `exit_code`
+            # property.
+            exit_codes = [
+                r['exit_code'] for r in exc.failed if 'exit_code' in r]
+            if exit_codes:
+                # we have exit code(s), take the last one
+                exit_code = exit_codes[-1]
         elif isinstance(exc, CommandError):
             exit_code = _communicate_commanderror(exc) or exit_code
         elif isinstance(exc, KeyboardInterrupt):
```

More precisely, it would introduce the behavior to relay any last exit code, whenever a command causes an `IncompleteResults` exception.

This seems like a broad change. However, this is also exactly what happens when a `CommandError` bubbles up to the main entry point. Its exit code is also relayed.

So this patch would only homogenize the behavior to also cover such exceptions, when they are wrapped into an error results.

**Comment by adswa** [comment]
I like this suggestion!

**Comment by yarikoptic** [comment]
sound sensible.  I don't know when we get multiple but I just wouldn't rely on exit_codes to be non-0s so I would return e.g. `max(exit_codes)`  or first explicitly non-0.  If all are 0s -- AssertionError of some kind since it would violate assumption code has

**Comment by yarikoptic-gitmate** [comment]
Issue fixed in [`1.1.4`](https://github.com/datalad/datalad/releases/tag/1.1.4)
